docs: add javascript section

happysalada opened the pull request:
###### Motivation for this change

This is a work in progress, but the aim to add a little more information for people trying to package things in javascript.

Here are some questions that remain to be answered.
- There is already a node.section, perhaps those 2 sections should be merged ? The node section aims to address how to add cli packages from npm, which is a slightly different problem.
- The sections on each of the individual tools are still incomplete. I'm wondering how people feel about those. If they are too detailed (referring specific github issues), or if they are doing a duplicate job that the upstream repo for the tool is already trying to do.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

happysalada commented on doc/languages-frameworks/javascript.section.md:
```diff
@@ -0,0 +1,113 @@
+
+Exceptions to this rule are
+
+- Sometimes some dependencies are assumed to global to the upstream repo. In that case you can add them manually to the upstream package.json (`yarn add xxx` or `npm install xxx`, ...). Dependencies that are installed locally can be executed with `npx` for cli tools. (e.g. `npx postcss ...`, this is how you can call those dependencies in the phases).
```
This is coming from https://github.com/arso-project/sonar 
they ask you to install a global dependency `@arsonar/server`
if you only install it in the local repo, you'll have to either change your `$PATH` to make it available or you can use it locally with `npx`.
I'm happy to change the wording to make it more understandable.

Sohalt replied:
```suggestion
- Sometimes the upstream repo assumes some dependencies to be installed globally. In that case you can add them manually to the upstream package.json (`yarn add xxx` or `npm install xxx`, ...). Dependencies that are installed locally can be executed with `npx` for cli tools. (e.g. `npx postcss ...`, this is how you can call those dependencies in the phases).
```

happysalada replied:
modified, thank you! Feel free to update in case you feel something is missing.

SuperSandro2000 commented:
I think you are missing that we prefer to use upstream tarballs because we do not need to commit a giant lock file.

jtojnar commented on doc/languages-frameworks/javascript.section.md:
```diff
@@ -0,0 +1,113 @@
+# Javascript
```
Would be nice to add anchors to other headings as well.

happysalada replied:
Added, let me know if you think it hasn't been properly addressed.

happysalada committed:
docs: add javascript section

happysalada commented:
I think I've addressed all the comments. I'm happy to change anything I might have forgotten.

I've added some more details regarding yarn2nix
I've added a section on packaging something outside of nixpkgs
I've added a commit moving the node section inside of the javascript section. I've done it in a separate commit to show i haven't modified anything from that section.

There are still things that could be improved, but I would like to submit this as a first pass.

happysalada committed:
docs: move node section to javascript section

billewanick commented on doc/languages-frameworks/javascript.section.md:
```diff
@@ -0,0 +1,203 @@
+
+#### mkYarnModules {#javascript-yarn2nix-mkYarnModules}
+
+this will generate a derivation including the node_modules. If you have to build a derivation for an integrated web framework (rails, phoenix..), this is probably the easiest way. Plausible offers a good example of how to do this.
```
Plausible is referenced here, but I'm not sure what that refers to.
Is it this repo? https://github.com/plausible/analytics

happysalada replied:
inside nixpkgs pkgs/servers/web-apps/plausible/default.nix I wasn't sure if I could do a file reference in markdown.

happysalada replied:
Ok, I added the two links to the plausible references.

billewanick replied:
Looks great, thank you!
